pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java

```diff
@@ -0,0 +1,273 @@
+ *
+ * When entire predicate gets rewritten to false (Example 3 above), the query will not return any data. Hence, it is
+ * better for the Broker itself to return an empty response rather than sending the query to servers for further
+ * evaluation.
+ */
+public class NumericalFilterOptimizer implements FilterOptimizer {
+
+  private static final Expression TRUE = RequestUtils.getLiteralExpression(true);
+  private static final Expression FALSE = RequestUtils.getLiteralExpression(false);
+
+  @Override
+  public FilterQueryTree optimize(FilterQueryTree filterQueryTree, @Nullable Schema schema) {
+    // Don't do anything here since this is for PQL queries which we no longer support.
+    return filterQueryTree;
+  }
+
+  @Override
+  public Expression optimize(Expression expression, @Nullable Schema schema) {
+    ExpressionType type = expression.getType();
+    if (type != ExpressionType.FUNCTION) {
+      // Not a function, so we have nothing to rewrite.
+      return expression;
+    }
+
+    Function function = expression.getFunctionCall();
+    List<Expression> operands = function.getOperands();
+    String operator = function.getOperator();
+    if (operator.equals(FilterKind.AND.name()) || operator.equals(FilterKind.OR.name())) {
+      // One of the operands may be an EQUALS or NOT_EQUALS function so recursively traverse the expression tree to see
+      // if we find an EQUALS or NOT_EQUALS function to rewrite.
+      operands.forEach(operand -> optimize(operand, schema));
+
+      // We have rewritten the child operands, so rewrite the parent if needed.
+      return optimizeCurrent(expression);
+    } else if (operator.equals(FilterKind.EQUALS.name()) || operator.equals(FilterKind.NOT_EQUALS.name())) {
+      // Verify that LHS is a numeric column and RHS is a numeric literal before rewriting.
+      Expression lhs = operands.get(0);
+      Expression rhs = operands.get(1);
+      if (isNumericColumn(lhs, schema) && isNumericLiteral(rhs)) {
+        // Rewrite the expression.
+        return rewrite(expression, lhs, rhs, schema);
+      }
+    }
+
+    return expression;
+  }
+
+  /**
+   * If any of the operands of AND function is "false", then the AND function itself is false and can be replaced with
+   * "false" literal. Otherwise, remove all the "true" operands of the AND function. Similarly, if any of the operands
+   * of OR function is "true", then the OR function itself is true and can be replaced with "true" literal. Otherwise,
+   * remove all the "false" operands of the OR function.
+   */
+  private static Expression optimizeCurrent(Expression expression) {
+    Function function = expression.getFunctionCall();
+    List<Expression> operands = function.getOperands();
+    if (function.getOperator().equals(FilterKind.AND.name())) {
+      // If any of the literal operands are FALSE, then replace AND function with FALSE.
+      if (operands.stream().anyMatch(operand -> operand.equals(FALSE))) {
```

Comment:
Let's avoid java8 functional programming functions as we run into performance issues before

Reply:
Fixed.